block-producer/src/block_builder/prover/mod.rs

```diff
@@ -57,6 +71,51 @@ begin
     drop
     # => [ROOT_{n-1}]
 end
+
+#! Compute the note root
+#! 
+#! Stack: [num_notes_updated, SMT_EMPTY_ROOT, note_key_0, NOTE_HASH_0, ... , note_key_{n-1}, NOTE_HASH_{n-1}]
+#! Output: [NOTES_ROOT]
+proc.compute_note_root
+    # assess if we should loop
+    dup neq.0 
+    #=> [0 or 1, num_notes_updated, SMT_EMPTY_ROOT, ... ]
+
+    while.true
+        #=> [notes_left_to_update, ROOT_i, note_key_i, NOTE_HASH_i, ... ]
+
+        # Prepare stack for mtree_set
+        movdn.9 movup.4 push.8
+        #=> [depth=8, note_key_i, ROOT_i, NOTE_HASH_i, notes_left_to_update, ... ]
+
+        mtree_set dropw 
+        #=> [ROOT_{i+1}, notes_left_to_update, ... ]
+
+        # loop counter
+        movup.4 sub.1 dup neq.0
+        #=> [0 or 1, notes_left_to_update - 1, ROOT_{i+1}, ... ]
+    end
+
+    drop
+    # => [ROOT_{n-1}]
+end
+
+# Stack: [<account root inputs>, <note root inputs>]
+proc.main.2
+    exec.compute_account_root loc_storew.0 dropw
+    #=> [<note root inputs>]
+
+    exec.compute_note_root loc_storew.1 dropw
+    #=> [ ]
+
+    # Load output on stack
+    loc_loadw.1 padw loc_loadw.0
+    #=> [ ACCOUNT_ROOT, NOTE_ROOT]
+end
```

Comment:
Is there a reason to have a dedicated `main` function rather than put the same code into `begin ... end` section? The only thing that would need to change is that we'd use global memory (i.e., via `mem_storew`) rather than local procedure memory.

Reply:
I liked to let the assembler figure out where to store my objects as opposed to me needing to define my own constants. Basically, I prefer reading `loc_storew.0` than `mem_storew.ACCOUNT_ROOT_PTR`, especially when used as one liners:

```
// less verbose
loc_loadw.1 padw loc_loadw.0
```
vs
```
// more verbose, but at the same time, a bit more clear...
mem_loadw.NOTE_ROOT_PTR padw mem_loadw.ACCOUNT_ROOT_PTR
```

This is not a strong preference though; I'm happy to change it to use global memory instead.

Reply:
We can use inline constants with `mem_storew` as well. Basically, to re-write the `main` procedure as `begin ... end` block, we'd need to do this:
```
begin
    exec.compute_account_root mem_storew.0 dropw
    #=> [<note root inputs>]

    exec.compute_note_root mem_storew.1 dropw
    #=> [ ]

    # Load output on stack
    mem_loadw.1 padw mem_loadw.0
    #=> [ ACCOUNT_ROOT, NOTE_ROOT]
end
```

Btw, a slightly more efficient way to write this would be:
```
begin
    exec.compute_account_root mem_storew.0 dropw
    #=> [<note root inputs>]

    exec.compute_note_root
    #=> [NOTE_ROOT]

    # Load output on stack
    padw mem_loadw.0
    #=> [ ACCOUNT_ROOT, NOTE_ROOT]
end
```

Reply:
> We can use inline constants with mem_storew as well.

ah I didn't know you could do that!

Fixed